LabExT/Movement/Transformations.py

```diff
@@ -597,7 +587,97 @@ def stage_to_chip(
+            self.translation_to_chip.T).flatten())
+
+
+def rigid_transform_with_orientation_preservation(
+    S: np.ndarray,
+    T: np.ndarray,
+    axes_rotation: np.ndarray = None
+) -> None:
+    """
+    Calculates a rotation matrix that provides optimal rotation with respect
+    to least squares error between two sets of 3D coordinates.
+
+    The algorithm tries to maintain the orientation of the unit vectors after rotation.
+    The axis rotation previously defined by the user is taken as the ground truth.
+    Each unit vector is applied from the matrix and the scalar point is used to determine
+    whether the unit vector points in the same direction after the axis rotation and after the rotation. If not, it is inverted.
+
+    Parameters
+    ----------
+    target_dataset : np.ndarray
+        3xN dataset of N 3-D coordinates (later referenced as T)
+    start_dataset : np.ndarray
+        3xN dataset of N-3D coordinates (later referenced as S)
+    axes_rotation : np.ndarray
+        3x3 ground truth axes rotation used for orientation preservation
+
+    The algorihm computes a matrix R and translation vector t, such that:
+        RS + t = T
+
+    More details: https://github.com/nghiaho12/rigid_transform_3D/blob/master/rigid_transform_3D.py
+    """
+    if axes_rotation is not None and axes_rotation.shape != (3, 3):
+        raise ValueError(
+            f"Axes rotation matrix must be 3x3, got shape {axes_rotation.shape}")
+
+    if not S.shape == T.shape:
+        raise ValueError("Start and target dataset must have the same shape")
+
+    target_num_rows, target_num_cols = T.shape
+    start_num_rows, start_num_cols = S.shape
+
+    if target_num_rows != 3:
+        raise ValueError(
+            f"Target dataset is not Nx3, got shape {target_num_rows}x{target_num_cols}")
+
+    if start_num_rows != 3:
+        raise ValueError(
+            f"Start dataset is not Nx3, got shape {start_num_rows}x{start_num_cols}")
+
+    # Centroid col wise
+    t_centroid = np.mean(T, axis=1, keepdims=True)
+    s_centroid = np.mean(S, axis=1, keepdims=True)
+
+    # Subtract Centroid
+    T_c = T - t_centroid
+    S_c = S - s_centroid
+
+    # Calculate accumulating matrix H = ST^T
+    H = S_c @ np.transpose(T_c)
+
+    # Calculate SVD: [U, S, V] = SVD(H)
+    U, _, V = np.linalg.svd(H)
+
+    # Calculate R
+    R = V.T @ U.T
+
+    # special reflection case
+    if np.linalg.det(R) < 0:
+        print("det(R) < R, reflection detected!, correcting for it ...")
+        V[2, :] *= -1
+        R = V.T @ U.T
+
+    # autocorrect orientation
+    if axes_rotation is not None:
+        correction_matrix = np.identity(3)
+        for i, unit_vector in enumerate(
+                [np.array([1, 0, 0]), np.array([0, 1, 0]), np.array([0, 0, 1])]):
+            ground_truth = axes_rotation @ unit_vector
+            if ground_truth.dot(R @ unit_vector) < 0:
+                correction_matrix[i, i] = -1
+
+        R = R @ correction_matrix
```

Comment:
Will be interesting to see if this works with non Left-Right oriented stages? Do you test for this?

Reply:
Unfortunately not in simulation, because I don't have a calibration for top and bottom oriented stages.
The method is only tested for random axis rotations with random coordinate pairings.